src/metabase/driver/mysql.clj

```diff
@@ -348,6 +353,40 @@
                                 2)
                           (hx/literal "-01"))))
 
+(defmethod sql.qp/->honeysql [:mysql :datetime-diff]
+  [driver [_ x y unit :as clause]]
+  (let [x (sql.qp/->honeysql driver x)
+        y (sql.qp/->honeysql driver y)
+        disallowed-types (keep
+                          (fn [v]
+                            (when-let [db-type (some-> v hx/type-info hx/type-info->db-type keyword str/upper-case)]
+                              (let [base-type (sql-jdbc.sync/database-type->base-type driver db-type)]
+                                (when-not (some #(isa? base-type %) [:type/Date :type/DateTime])
+                                  (name db-type)))))
+                          [x y])]
+    (when (seq disallowed-types)
+      (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                           (pr-str disallowed-types))
+                      {:found disallowed-types
+                       :type  qp.error-type/invalid-parameter})))
+    (case unit
+      (:year :month)
+      (hsql/call :timestampdiff (hsql/raw (name unit)) (hsql/call :date x) (hsql/call :date y))
+
+      :week
+      (let [positive-diff (fn [a b] (hx/floor (hx// (hsql/call :datediff b a) 7)))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :day
+      (hsql/call :datediff y x)
+
+      (:hour :minute :second)
+      (hsql/call :timestampdiff (hsql/raw (name unit)) x y)
+
+      (throw (ex-info (tru "Invalid datetime-diff unit {0}" unit)
+                      {:clause      clause
+                       :valid-units [:year :month :week :day :hour :minute :second]
+                       :type        qp.error-type/invalid-parameter})))))
```

Comment:
This is not something that should be happening in each individual driver. If it's in the MBQL schema then the validate-query middleware, which happens right after normalization, will reject it. 

In general when working on QP/driver stuff stop and ask yourself "is this something that can be done once generally, or is it something that has to be done 20 times for 20 drivers?". The only things that need to be done for each individual driver are things that are driver-specific, like how to translate a specific MBQL clause to their flavor of SQL/native query language. MBQL transformations and validation are not driver-specific and should not be done in each individual driver.

This doesn't really apply to the type validation at the beginning of this method since column type info is DB-dependent. Maybe in the future we can have better typing around MBQL stuff and some of that can be determined by middleware before it ever hits the QP but for the time being it's ok

Reply:
Yes, this is actually caught by the MBQL schema anyway. I've removed this error handling in https://github.com/metabase/metabase/commit/e2a8f3be32855048d1310163991e4dd0b919f0cb 

In theory the `unit` argument should be allowed to be a string expression or field, and if so, we'd have to bring back this error handling to each driver, right? Either way, I'd leave it for the future.